### ISSUE TITLE
Module level, inline assembly for wasm segfaults.

### ISSUE THREAD

**Issue by gcoakes** [ziglang/zig#13129]
### Zig Version

0.10.0-dev.4254+7f508480f

### Steps to Reproduce

main.zig:

``` zig
comptime {
    asm (
        \\.globl wasm_nop
        \\wasm_nop:
        \\  .functype wasm_nop () -> ()
        \\  nop
        \\  end_function
    );
}
```

Compile:

```
$ zig build-lib -dynamic -target wasm32-freestanding main.zig
LLVM Emit Object... Segmentation fault at address 0x0
???:?:?: 0x7f2d2771d249 in ??? (???)
???:?:?: 0x72732f73656b616e in ??? (???)
fish: Job 1, 'zig build-lib -dynamic -target…' terminated by signal SIGABRT (Abort)
```

### Expected Behavior

A wasm file at least containing:

``` wat
(module (func $wasm_nop nop))
```

### Actual Behavior

```
$ zig build-lib -dynamic -target wasm32-freestanding main.zig
LLVM Emit Object... Segmentation fault at address 0x0
???:?:?: 0x7f2d2771d249 in ??? (???)
???:?:?: 0x72732f73656b616e in ??? (???)
fish: Job 1, 'zig build-lib -dynamic -target…' terminated by signal SIGABRT (Abort)
```

**Comment by nektro** [comment]
```zig
export fn wasm_nop() void {}
```

this is the way to do that. using `asm` inside a `comptime` block is illegal. however, this issue should stay open because that should be a compiler error not a compiler panic

**Comment by gcoakes** [comment]
> using asm inside a comptime block is illegal

I don't think this is generally correct. There are [behavior tests](https://github.com/ziglang/zig/blob/c3d67c5c4e2e86b472ae046f5e0e39db61009213/test/behavior/asm.zig#L7-L30) for this, and #256 added support for it. Are you saying it's just not legal for WASM?

Also, my use case is not strictly the minimal example I gave. I want to be able to define wrapper functions (a la [wasm-bindgen](https://rustwasm.github.io/docs/wasm-bindgen/contributing/design/js-objects-in-rust.html)) with parameters which are not expressible in zig (#10491).

**Comment by gcoakes** [comment]
Slightly off topic, but I tried building the failing case using a debug build of LLVM as described in the [wiki](https://github.com/ziglang/zig/wiki/How-to-build-LLVM,-libclang,-and-liblld-from-source#debug) (which took ages). It produced identical output to my original report. @Luukdegram, was there something different you used to produce your stack trace?

**Comment by Luukdegram** [comment]
> Slightly off topic, but I tried building the failing case using a debug build of LLVM as described in the [wiki](https://github.com/ziglang/zig/wiki/How-to-build-LLVM,-libclang,-and-liblld-from-source#debug) (which took ages). It produced identical output to my original report. @Luukdegram, was there something different you used to produce your stack trace?

Did you also build a debug version of the Zig compiler and not only of LLVM?

**Comment by gcoakes** [comment]
It's not *all* module-level asm. Maybe something specific to defining a function there? The following compiles:

```zig
comptime {
    asm (
        \\.globl __refs
        \\.tabletype __refs, externref, 32
        \\__refs:
    );
}

export fn refsSize() usize {
    return asm volatile (
        \\table.size __refs
        \\local.set %[ret]
        : [ret] "=r" (-> usize),
    );
}
```

**Comment by alexrp** [comment]
This seems fixed. But note https://github.com/ziglang/zig/issues/24077.
